jlm/llvm/opt/alias-analyses/Steensgaard.cpp

```diff
@@ -1379,45 +1379,69 @@ Steensgaard::AnalyzeMemcpy(const jlm::rvsdg::simple_node & node)
+  //
+  // The basic idea behind memcpy is that the memory where src is pointing to is copied(!)
+  // to dst. This means that src and dst do not point to the same memory in the end, but the
+  // points-to sets from both need to point to the same memory. The result is the following graph:
+  //
+  // src -> srcMemory
+  // dst -> dstMemory (which is a copy of srcMemory)
+  // srcMemory and dstMemory -> underlyingMemory
+  //
+  // It might be that we do not know the MemoryLocation(s) to which srcMemory, dstMemory, or
+  // underlyingMemory is pointing to yet(!), and we might never know. We therefore insert dummy
+  // locations (if necessary) such that we can express the above relationships.
   if (srcAddress.GetPointsTo() == nullptr)
   {
-    // If we do not know where the source address points to yet(!),
-    // insert a dummy location such that we have something to work with.
     auto & dummyLocation = LocationSet_->InsertDummyLocation();
     srcAddress.SetPointsTo(dummyLocation);
   }
 
   if (dstAddress.GetPointsTo() == nullptr)
   {
-    // If we do not know where the destination address points to yet(!),
-    // insert a dummy location such that we have something to work with.
     auto & dummyLocation = LocationSet_->InsertDummyLocation();
     dstAddress.SetPointsTo(dummyLocation);
   }
 
   auto & srcMemory = LocationSet_->GetRootLocation(*srcAddress.GetPointsTo());
   auto & dstMemory = LocationSet_->GetRootLocation(*dstAddress.GetPointsTo());
 
-  if (srcMemory.GetPointsTo() == nullptr)
+  if (srcMemory.GetPointsTo() == nullptr && dstMemory.GetPointsTo() == nullptr)
   {
-    auto & dummyLocation = LocationSet_->InsertDummyLocation();
-    srcMemory.SetPointsTo(dummyLocation);
+    auto & underlyingMemory = LocationSet_->InsertDummyLocation();
+    srcMemory.SetPointsTo(underlyingMemory);
+    dstMemory.SetPointsTo(underlyingMemory);
   }
-
-  if (dstMemory.GetPointsTo() == nullptr)
+  else
   {
-    auto & dummyLocation = LocationSet_->InsertDummyLocation();
-    dstMemory.SetPointsTo(dummyLocation);
+    if (srcMemory.GetPointsTo() == nullptr)
+    {
+      auto & dummyLocation = LocationSet_->InsertDummyLocation();
+      srcMemory.SetPointsTo(dummyLocation);
+    }
+
+    if (dstMemory.GetPointsTo() == nullptr)
+    {
+      auto & dummyLocation = LocationSet_->InsertDummyLocation();
+      dstMemory.SetPointsTo(dummyLocation);
+    }
+
+    // Unifies the underlying memory of srcMemory and dstMemory
+    LocationSet_->Join(*srcMemory.GetPointsTo(), *dstMemory.GetPointsTo());
   }
 
-  LocationSet_->Join(*srcMemory.GetPointsTo(), *dstMemory.GetPointsTo());
+  // Handle flags
+  //
+  // The dstMemory needs to have at least the same flags as the src address, because memcpy copies
+  // over the memory from src to dst. Thus, if src is flagged as pointing to escaped memory, than
+  // dstMemory needs also to be flagged as pointing to escaped memory as well.
+  auto dstMemoryPointsToFlags = dstAddress.GetPointsToFlags();
```

Comment:
Yikes. Good catch. I meant `srcAddress` and `dstMemory`, not `dstAddress`.

Reply:
So what we have currently is our source pointer `s` and our destination pointer `d`. After I handled the points-to sets in `AnalyzeMemcpy`, I have something like this: `s` -> `x` -> `u` and `d` -> `y` -> `u`, where `y` is a copy of `x` and both, `x` and `y`, point to `u`, i.e., `underlyingMemory`. Now, I have the problem that `s` might also point to escaped/external memory, i.e., it has these flags. That automatically means that `x` points to escaped/external as well and with that also `u` (even though they might not be annotated yet, as the propagate pass has not run yet). Since `y` is a copy of `x`, it would also mean that `y` needs to be marked. That is why I assign the flags from `srcAddress`, i.e., `s`, to `y`.

Reply:
Ah, I see. If s finds out about pointing to escaped later, it might still be an issue? But within the current flag system this makes sense

Reply:
I just had another round at this and realized that it does not work like I wanted it to. Simple counter example:

```
extern int* global;

static void
g(int* s)
{
  int localInt;
  memcpy(&localInt, s, sizeof(int));
}

void
f()
{
  g(global);
}
```

The functions are processed top-down in the RVSDG. This means `g()` is analyzed before `f()`. Function `g()` is static, so its argument is not marked as escaped/external. Thus, when it analyzes the memcpy, `srcAddress` will not (yet) be marked as escaped, and therefore dstMemory will not be marked as escaped. Once it analyzes function `f()`, it will mark the argument of `g()` as escaped (by handling the call to `g()`), but then we already handled memcpy(). So we lost.

I (potentially) see two ways out of this:
1. Start to analyzes the graph bottom-up. This would ensure that we always handle the call sites before the actual function (some special handling would need to be necessary for recursive functions in phi environments).
2. We do not separate between srcMemory and dstMemory and simply treat them as a single location, i.e., we do not treat it as a copy and simply let srcAddress and dstAddress point to the same memory location.

I am a little bit reluctant in going for option 1 as I do not fully see the overall consequences of it (maybe there are some hidden details I am not thinking about right now and it actually won't work) . Also it would mean quite a big restructuring of the entire analysis.

Option 2 has the downside of losing some precision. However, thinking about it, a memcpy is nothing else than a load followed by a store. Looking at how this is modeled in Steensgaard, then we would never have had a `srcMemory` and `dstMemory` in the first place, but both, `srcAddress` and `dstAddress` pointing to the same memory location. Thus, the perceived precision might not have been there in the first place.

I think I will go for option 2. Option 1 (if it is even a real option) would be a bigger process that I do not want to invest into.

Thanks for being so persistent. I appreciate it!